### ISSUE TITLE
Misleading diagnostic for macro `error!` when called with no parameters

### ISSUE THREAD

**Issue by MarkSteinbrick95** [rust-lang/rust#58796]
When calling the macro with no arguments `error!()`:
```rust
error: Could not compile
warning: build failed, waiting for other jobs to finish...
error: expected identifier, found `,`
  --> mod.rs:37:15
   |
37 |               error!();
   |               ^^^^^^^^^
   |               |
   |               expected identifier
   |               help: remove this comma
   |
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error: aborting due to previous error

error: Could not compile
```

The diagnostic should instruct the user that a string literal is required as a parameter to the macro.

**Comment by jonas-schievink** [comment]
> The diagnostic should instruct the user that a string literal is required as a parameter to the macro.

`error!` is declared as taking `$($arg:tt)*`, so I don't think this is possible. If it would take a `$fmt:lit` instead, maybe.

**Comment by estebank** [comment]
Ran using `-Zexternal-macro-backtrace`:

```
error: expected identifier, found `,`
  --> <::log::macros::error macros>:4:28
   |
1  | / ( target : $ target : expr , $ ( $ arg : tt ) * ) => (
2  | | log ! ( target : $ target , $ crate :: Level :: Error , $ ( $ arg ) * ) ; ) ;
3  | | ( $ ( $ arg : tt ) * ) => (
4  | | log ! ( $ crate :: Level :: Error , $ ( $ arg ) * ) ; )
   | |____________________________^__________________________- in this expansion of `error!`
   |                              |
   |                              expected identifier
   |                              help: remove this comma
   |
  ::: src/main.rs:29:5
   |
29 |       error!();
   |       --------- in this macro invocation
```

Getting the compiler to behave correctly in this case will be difficult, but making the macro itself handle this case and have [an appropriate compile time error](https://doc.rust-lang.org/nightly/std/macro.compile_error.html) might be reasonable. What does @rust-lang/libs think?

**Comment by dtolnay** [comment]
I opened https://github.com/rust-lang-nursery/log/pull/322 to fix the problem on log's side. The compiler was largely doing the right thing and I consider it a problem with the way log's macro was written that resulted in a misleading error.

There is something going wrong in the compiler here too though. Minimized:

```rust
macro_rules! log {
    ($lvl:expr, $($arg:tt)+) => {}
}

fn main() {
    log!(Level::Error ,);
}
```

```console
error: expected identifier, found `,`
 --> src/main.rs:6:17
  |
6 |     log!(Level::Error ,);
  |                 ^
  |                 |
  |                 expected identifier
  |                 help: remove this comma
```

The error points to something that is an identifier and says it expected an identifier :disappointed: and points to something that is not a comma and says remove this comma. :cry:

**Comment by estebank** [comment]
I noticed that too. My guess is that it's just a token off by one error somewhere.

**Comment by estebank** [comment]
@dtolnay I can confirm that that incorrect diagnostic is caused by https://github.com/rust-lang/rust/pull/52397 :(
